src/hypothesis/strategies.py

```diff
@@ -1575,4 +1578,96 @@ def deferred(definition):
     return DeferredStrategy(definition)
 
 
+class _PathLike(object):
+    """A class implementing the os.PathLike protocol available since Python
+    3.6.
+
+    See https://www.python.org/dev/peps/pep-0519/ for more details.
+    """
+
+    def __init__(self, value):
+        self._value = value
+
+    def __fspath__(self):
+        return self._value
+
+    def __repr__(self):
+        return 'pathlike(%r)' % self._value
+
+
+@defines_strategy
+@composite
+def fspaths(draw, allow_pathlike=True, allow_existing=False):
+    """A strategy which generates filesystem path values.
+
+    The generated values include everything which the builtin
+    :func:`python:open` function accepts i.e. which won't lead to
+    :exc:`ValueError` or :exc:`TypeError` being raised.
+
+    Note that the range of the returned values depends on the operating
+    system, the Python version, and on the filesystem encoding as returned by
+    :func:`sys.getfilesystemencoding`.
+
+    :param bool allow_pathlike:
+        If the result can be a pathlike (see :class:`os.PathLike`)
+    :param bool allow_existing:
+        If paths which happen to exist on the filesystem should be returned.
+        This is :obj:`python:False` by default to prevent tests from accessing
+        existing files by accident.
+
+    .. versionadded:: 3.15
+
+    """
+
+    strategies = []
+
+    if os.name == 'nt':  # pragma: no cover
+        hight_surrogate = characters(
+            min_codepoint=0xD800, max_codepoint=0xDBFF)
+        low_surrogate = characters(
+            min_codepoint=0xDC00, max_codepoint=0xDFFF)
+        uni_char = characters(min_codepoint=0x1)
+        windows_path_text = text(
+            alphabet=one_of(uni_char, hight_surrogate, low_surrogate))
+        strategies.append(windows_path_text)
+
+        def text_to_bytes(path):
+            fs_enc = sys.getfilesystemencoding()
+            try:
+                return path.encode(fs_enc, 'surrogatepass')
+            except UnicodeEncodeError:
+                return path.encode(fs_enc, 'replace')
+
+        windows_path_bytes = windows_path_text.map(text_to_bytes)
+        strategies.append(windows_path_bytes)
+    else:
+        unix_path_bytes = binary().map(lambda b: b.replace(b"\x00", b" "))
+        strategies.append(unix_path_bytes)
+
+        unix_path_text = unix_path_bytes.map(
+            lambda b: b.decode(
+                sys.getfilesystemencoding(),
+                'surrogateescape' if PY3 else 'ignore'))
+
+        random = draw(randoms())
+
+        def shuffle_text(t):
+            l = list(t)
+            random.shuffle(l)
+            return u"".join(l)
+
+        strategies.append(unix_path_text.map(shuffle_text))
+
+    main_strategy = one_of(strategies)
```

Comment:
For my library my motivation was mostly covering all value/type variants and less about the path itself, thus the limited output.

This all sounds good. I'll try to come up with something.